scipy/_lib/_util.py

```diff
@@ -153,9 +153,9 @@ def _lazywhere(cond, arrays, f, fillvalue=None, f2=None):
         temp2 = xp.asarray(f2(*(arr[ncond] for arr in arrays)))
         dtype = xp.result_type(temp1, temp2)
         out = xp.empty(cond.shape, dtype=dtype)
-        out[ncond] = temp2
+        out = xp_at(out, ncond).set(temp2, copy=None)
 
-    out[cond] = temp1
+    out = xp_at(out, cond).set(temp1, copy=None)
```

Comment:
This won't work under `@jax.jit` because jax doesn't support unknown shapes.
There are two ways forward:
1. (easy-ish?) Reimplement this function in JAX or in array-api-extra and use two different code paths if it is being jitted or not. (_does such a flag exist today?_)
If jitted, use `out = jnp.where(cond, f(arr), f2(arr))`. _I think_ that the `@jax.jit` will cause `f` not to apply at all to `arr[~cond]` and `f2` not to apply to `arr[cond]`? 
2. (hard) change `@jax.jit` to track unknown shapes:
```python
UNKNOWN_SHAPE_0 = arr[cond].shape  # line 137 
temp1.shape = UNKNOWN_SHAPE_0  # line 137 
UNKNOWN_SHAPE_1 = arr[ncond].shape  # line 153
temp2.shape = UNKNOWN_SHAPE_1  # line 153
out = out.at[ncond].set(temp2)  # line 156; out[ncond] and temp2 have the same shape UNKNOWN_SHAPE_1
out = out.at[cond].set(temp1)  # line 158; out[cond] and temp1 have the same shape UNKNOWN_SHAPE_0
```
In other words, this would require `@jax.jit` to support unknown shapes _as long as they are created and reabsorbed within the same jitted code_.

Reply:
CC @rgommers @jakevdp @pearu

Reply:
As a general comment from 30000 feet (am not reviewing this specific PR and do not know enough about jax internals to say anything about specifics):

if a body of this sort of backend-specific workarounds grows beyond a handful of functions to put into array-api-extra, I would suggest to seriously consider using a delegation story [1]. These jax-delegating functions may either live in jax itself or in a separate package, as decided by jax devs and scipy collaborators invested in jax support. Making scipy functions know if they are being jax-jitted or not is just a bit too much. Same for torch.compile or any other existing or future jit frameworks and, in fact, generally the majority of backend-specific code.


[1] For cupy, we make, for instance, `scipy.ndimage.<function>` _delegate_ to `cupyx.scipy.ndimage.<function>` if inputs are cupy arrays. Ditto for `jax.scipy` functions.

Reply:
CC @vfdev-5

Reply:
> * (easy-ish?) Reimplement this function in JAX or in array-api-extra and use two different code paths if it is being jitted or not. (_does such a flag exist today?_)

There is no supported way in JAX to check if code is under JIT.

>   If jitted, use `out = jnp.where(cond, f(arr), f2(arr))`. _I think_ that the `@jax.jit` will cause `f` not to apply at all to `arr[~cond]` and `f2` not to apply to `arr[cond]`?

XLA will generally execute both `f` and `f2` on the entire array. This is because on accelerators, there is no efficient way to apply a function to only part of an array. Nevertheless, using `where` would be the recommended way to write this operation in JAX.

> * (hard) change `@jax.jit` to track unknown shapes:

LOL

Reply:
> There is no supported API in JAX to check if code is under JIT

@vfdev-5 suggested offline:
> if the functions is jitted, then all array-like structures are replaced with a Tracer objects so to know if you are in the eager mode or not, you can do as here: https://github.com/jax-ml/jax/blob/a123d4e39ee67585c0a9dde4694db38ba1aa3e9a/jax/_src/numpy/lax_numpy.py#L6808-L6809

does your comment mean that such a test is disencouraged because it requires the user to test two separate code paths, or that it won't work in some cases?

> LOL

Could you expand? Does it mean "LOL, that's nightmarishly hard don't even bother trying" or "LOL, we've been working on this for ages and we're on track to deliver it already"?

Reply:
> on accelerators, there is no efficient way to apply a function to only part of an array at dynamically-specified locations.

...from my memory of CUDA 101, you can apply `y = cond ? f(x) : f2(x)` but with the caveat that if cond has disjoint values within the warp (32 contiguous points) you _may_ end up waiting for both f1 and f2?
XREF https://stackoverflow.com/questions/29896422/cuda-avoiding-serial-execution-on-branch-divergence

Reply:
> does your comment mean that such a test is disencouraged because it requires the user to test two separate code paths, or that it won't work in some cases?

There is no public API to check whether you're under JIT, because doing so is a very easy way to inadvertently introduce subtle bugs into your code. It is of course possible to use non-public implementation details to infer whether you're in JIT, and that might work until those implementation details change. That's not a path I'd suggest for a project as important as SciPy.

> from my memory of CUDA 101, you can apply `y = cond ? f1(x) : f2(x)` but with the caveat that if cond has disjoint values within the warp (32 contiguous points) you _may_ end up waiting for both f1 and f2?

Yes, that's the case for a scalar `cond`, and JAX provides this API via `jax.lax.cond`. The context here if I understand correctly is a vector `cond`, which is a different situation.

> Could you expand?

Sorry, LOL probably wasn't that helpful. There have been several attempts at introducing some form of dynamic shapes into JAX over the years, but all have run into fundamental difficulties, not least of which is the non-existence of a suitable dynamic-shape-aware compiler. It may happen some day, but "hard" is an understatement 😁

Reply:
> Yes, that's the case for a scalar cond, and JAX provides this API via [jax.lax.cond](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.cond.html). The context here if I understand correctly is a vector cond, which is a different situation.

My snippet was in the context of an embarassingly-parallel CUDA kernel, where x, y, and cond are single points out of vectors

Reply:
> There is no public API to check whether you're under JIT, because doing so is a very easy way to inadvertently introduce subtle bugs into your code. It is of course possible to use non-public implementation details to infer whether you're in JIT, and that might work until those implementation details change. That's not a path I'd suggest for a project as important as SciPy.

Would you agree that `_lazy_where` (I dislike the name though; I think `apply_where` would be better) is a very generic and powerful function and it should be implemented in JAX with two separate code paths for jit vs. non-jit, plus a wrapper in `array-api-extra` to cover all other array libraries?

Reply:
My understanding is that embarrassingly parallel in CUDA implies vectorization. In JAX that’s certainly the case

Reply:
Alternatively, we could expand `jnp.at[ ].apply()` to work with any jit-table callable instead of just ufuncs?